clippy_lints/src/unnecessary_reserve.rs

```diff
@@ -0,0 +1,140 @@
+    pub fn new(conf: &'static Conf) -> Self {
+        Self {
+            msrv: conf.msrv.clone(),
+        }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryReserve {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        if !self.msrv.meets(msrvs::EXTEND_IMPLICIT_RESERVE) {
+            return;
+        }
+
+        if let ExprKind::MethodCall(PathSegment { ident: method, .. }, struct_calling_on, args_a, _) = expr.kind
+            && method.name.as_str() == "reserve"
+            && acceptable_type(cx, struct_calling_on)
+            && let Some(block) = get_enclosing_block(cx, expr.hir_id)
+            && let Some(next_stmt_span) = check_extend_method(cx, block, struct_calling_on, &args_a[0])
+            && !next_stmt_span.from_expansion()
+        {
+            let stmt_span = cx
+                .tcx
+                .hir()
+                .parent_id_iter(expr.hir_id)
+                .next()
+                .map_or(expr.span, |parent| cx.tcx.hir().span(parent));
+
+            span_lint_and_then(
+                cx,
+                UNNECESSARY_RESERVE,
+                next_stmt_span,
+                "unnecessary call to `reserve`",
+                |diag| {
+                    diag.span_suggestion(
+                        stmt_span,
+                        "remove this line",
+                        String::new(),
+                        Applicability::MaybeIncorrect,
+                    );
+                },
+            );
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+fn acceptable_type(cx: &LateContext<'_>, struct_calling_on: &Expr<'_>) -> bool {
+    if let Some(did) = adt_def_id(cx.typeck_results().expr_ty_adjusted(struct_calling_on)) {
+        matches!(cx.tcx.get_diagnostic_name(did), Some(sym::Vec | sym::VecDeque))
+    } else {
+        false
+    }
+}
+
+#[must_use]
+fn check_extend_method<'tcx>(
+    cx: &LateContext<'tcx>,
+    block: &'tcx Block<'tcx>,
+    struct_expr: &Expr<'tcx>,
+    args_a: &Expr<'tcx>,
+) -> Option<rustc_span::Span> {
+    let mut found_reserve = false;
+    let mut read_found = false;
+    let mut spanless_eq = SpanlessEq::new(cx);
+
+    let _: Option<!> = for_each_expr(cx, block, |expr: &Expr<'tcx>| {
+        if !found_reserve {
+            if expr.hir_id == args_a.hir_id {
+                found_reserve = true;
+            }
+            return ControlFlow::Continue(());
+        }
+
+        if let ExprKind::MethodCall(_, struct_calling_on, _, _) = expr.kind
+            && let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
+            && let ExprKind::MethodCall(
+                PathSegment {
+                    ident: method_call_a, ..
+                },
+                ..,
+            ) = args_a.kind
+            && method_call_a.name == sym::len
+            && match_def_path(cx, expr_def_id, &paths::ITER_EXTEND)
+            && acceptable_type(cx, struct_calling_on)
```

Comment:
Why do you need to check for the type if you will compare the expressions anyway? Not doing so would even let you inline the shorter form of `acceptable_type()` at the only place where it would now be called.